File: src/com/venus/Tree/BBSTree.java
package com.venus.Tree;

import java.util.Comparator;

public class BBSTree<E> extends BSTree<E>{

    public BBSTree(){
        this(null);
    }

    public BBSTree(Comparator<E> comparator) {
        super(comparator);
    }

    protected void rotateRight(Node<E> grand) {
        Node<E> parent = grand.left;
        Node<E> child = parent.right;

        grand.left = child;
        parent.right = grand;

        afterRotate(grand, parent, child);
    }

    protected void rotateLeft(Node<E> grand) {
        Node<E> parent = grand.right;
        Node<E> child = parent.left;

        grand.right = child;
        parent.left = grand;

        afterRotate(grand, parent, child);
    }

    protected void afterRotate(Node<E> grand, Node<E> parent, Node<E> child) {
        parent.parent = grand.parent;
        if (grand.isLeftChild()) {
            grand.parent.left = parent;
        }else if (grand.isRightChild()){
            grand.parent.right = parent;
        }else {
            root = parent;
        }

        if (child != null) {
            child.parent = grand;
        }
        grand.parent = parent;

    }

    protected void rotate(Node<E> r, Node<E> a, Node<E> b, Node<E> c, Node<E> d, Node<E> e, Node<E> f, Node<E>g) {

        d.parent = r.parent;
        if (r.isLeftChild()) {
            r.parent.left = d;
        }else if(r.isRightChild()) {
            r.parent.right = d;
        }else {
            root = d;
        }

        b.left = a;
        b.right = c;
        if (c != null) {
            c.parent = b;
        }
        if (a != null) {
            a.parent = b;
        }


        f.left = e;
        f.right = g;
        if (e != null) {
            e.parent = f;
        }
        if (g != null) {
            g.parent = f;
        }

        d.left = b;
        d.right = f;
        b.parent = d;
        f.parent = d;

    }
}

File: Python/Tree/bbstree.py
## written by xiongbiao
## date 2020-5-20

from Tree.bstree import BinarySearchTree

class BalancedBinarySearchTree(BinarySearchTree):
    def __init__(self):
        super().__init__()

    '''
        右旋：找到grand的左节点parent和parent的右节点child，更改parent，child，更新grand的父节点的左右子节点
              更新grand和parent节点的高度
                  g                      p 
                /   \                  /   \
               p    t3      --->      n     g
             /  \                    / \   / \
            n    t2                 t0 t1 t2 t3
           / \
          t0  t1
          g为不平衡的节点，g左子树高度 - g右子树高度 > 1
        '''

    def rotate_right(self, grand):
        parent = grand.left
        child = parent.right

        grand.left = child
        parent.right = grand

        self.after_rotate(grand, parent, child)

    '''
    左旋：与右旋对称
    '''

    def rotate_left(self, grand):
        parent = grand.right
        child = parent.left

        grand.right = child
        parent.left = grand

        self.after_rotate(grand, parent, child)

    def after_rotate(self, grand, parent, child):
        parent.parent = grand.parent
        if grand.is_left_child():
            grand.parent.left = parent
        elif grand.is_right_child():
            grand.parent.right = parent
        else:
            self._root = parent

        if child:
            child.parent = grand
        grand.parent = parent

